### PR TITLE
"implemented filters and made changes to testing"

### PR DESCRIPTION
**Pull request by jhcoll**

### CODE REVIEW

File: src/main/java/com/scottlogic/filters/AuthorPostFilter.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public AuthorPostFilter(String authorInput) {
+        Author = authorInput;
+    }
+
+    @Override
+    public List<UserPost> filter(List<UserPost> inputList) {
+        if (inputList == null) {
+            return null;
+        }
+        if(Objects.equals(Author, "")){return new ArrayList<>();}
```
Comment (jluxford-scottlogic) [inline src/main/java/com/scottlogic/filters/AuthorPostFilter.java:23]: This is a reasonable way of handling the empty string, the only slight issue with it is that it is inconsistent and not immediately obvious whether this is going to return all of the list or just an empty list. This means that in some designs you wouldn't be able to make a filter like this and an error would be thrown in the constructor instead. Both approaches have their pros and cons and this approach is perfectly reasonable just thought I should mention the other ways that Java sometimes does things

File: src/main/java/com/scottlogic/filters/AuthorPostFilter.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import com.scottlogic.PostFilter;
+import com.scottlogic.UserPost;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class AuthorPostFilter implements PostFilter {
+
+    final String Author;
```
Comment (jluxford-scottlogic) [inline src/main/java/com/scottlogic/filters/AuthorPostFilter.java:12]: `final` here doesn't set a scope this means that I think (though I could be wrong) that this is a protected scoped variable which means that it can be access within the same module. In general it is preferred to put a `private` as well as a `final` to make the scoping explicit to save confusion.

File: src/test/java/com/scottlogic/filters/AuthorPostFilterTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        List<UserPost> inputList = Collections.emptyList();
+        List<UserPost> expected = Collections.emptyList();
+
+        List<UserPost> actual = new AuthorPostFilter("").filter(inputList);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void filter_1ItemList_1ItemList() {
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/filters/AuthorPostFilterTest.java:46]: Could be good to do a singular item that gets filtered out completely? Although it is covered by the 3 items to none it is at a different case level, this being the base case, so it could be worth adding it in despite this.

File: src/test/java/com/scottlogic/filters/AuthorPostFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        List<UserPost> expected = List.of(userPost1);
+
+        List<UserPost> actual = new AuthorPostFilter("Jane").filter(inputList);
+
+        assertEquals(expected, actual);
+    }
+
+
+    @Test
+    void filter_3Items2Match_2Items() {
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/filters/AuthorPostFilterTest.java:57]: Similarly to the comment above maybe a 3 items to 3 items case could be of use as 3 items looks to be your inductive case and so covering the basis might still be useful?

File: src/test/java/com/scottlogic/filters/AuthorPostFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        assertEquals(expected, actual);
+    }
+
+
+    @Test
+    void filter_3Items0Match_EmptyList() {
+        List<UserPost> inputList = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> expected = Collections.emptyList();
+
+        List<UserPost> actual = new AuthorPostFilter("").filter(inputList);
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/filters/AuthorPostFilterTest.java:82]: The empty list case is more of an edge case rather than the standard 0 matches case, thus I might prefer it if this were called `filter_3ItemsEmptyFilter_EmptyList` and then a `filter_3Items0match_EmptyList()` used a different filter

File: src/test/java/com/scottlogic/filters/KeywordContentPostFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class KeywordContentPostFilterTest {
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/filters/KeywordContentPostFilterTest.java:14]: In general similar comments to author

File: src/main/java/com/scottlogic/filters/KeywordContentPostFilter.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
+package com.scottlogic.filters;
+
+import com.scottlogic.PostFilter;
+import com.scottlogic.UserPost;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class KeywordContentPostFilter implements PostFilter {
```
Comment (jluxford-scottlogic) [inline src/main/java/com/scottlogic/filters/KeywordContentPostFilter.java:10]: In general similar comments to author

**Review by jluxford-scottlogic — APPROVED** [review]
Nice way of doing the filters, looking really good so far. Just a few comments on scoping of variables, different approaches of edge cases and the dreaded possible extra tests, but overall looking like really good work well done.

File: src/main/java/com/scottlogic/filters/AuthorPostFilter.java (hunk 1 above)
Comment (jhcoll) [reply src/main/java/com/scottlogic/filters/AuthorPostFilter.java:23]: Yeah that's useful, thanks
